proposals/csharp-11.0/low-level-struct-improvements.md

```diff
@@ -869,6 +869,193 @@ ref struct S<T>
+- `$local`: this is the lifetime of any value that exists on the method stack. It's effectively a name place holder for *current method*. It is implicitly defined in methods and can appear in method signatures except for any output position.
+- `$ro`: name place holder for the *return only* safe to escape scope
+- `$cm`: name place holder for the *calling method* safe to escape scope
+
+There are a few predefined relationships between lifetimes:
+
+- `lifetime $heap >= $a` for all lifetimes `$a`
+- `lifetime $cm >= $ro` 
+- `lifetime $x >= $local` for all predefined lifetimes. User defined lifetimes have no relationship to local unless explictly defined.
+
+Lifetime variables when defined on types can be invariant or covariant. These are expressed using the same syntax as generic parameters:
+
+```csharp
+// $this is covariant
+// $a is invariant
+ref struct S<out $this, $a> 
+```
+
+The lifetime of a ref is expressed by providing a lifetime argument to the ref. For example a `ref` that refers to the heap is expressed as `ref<$heap>`.
+
+When defining a constructor in the model the name `new` will be used for the method. It is necessary to have a parameter list for the returned value as well as the constructor arguments. This is necessary to express the relationship between constructor inputs and the constructed value. Rather than having `Span<$a><$ro>` the model will use `Span<$a> new<$ro>` instead. The type of `this` in the constructor, including lifetimes, will be the defined return value.
+
+The basic rules for the lifetime are defined as:
+
+- All lifetimes are expressed syntactically as generic arguments, coming before type arguments. This is true for predefined lifetimes except `$heap` and `$local`. 
+- All types `T` that are not a `ref struct` implicitly have lifetime of `T<$heap>`. This is implicit, there is no need to write `int<$heap>` in every sample.
+- For a ref field defined as `ref T<$l1, $l2, ... $ln>` all lifetimes `$l1` through `$ln` must be invariant. 
+- For a ref defined as `ref<$a> T<$b, ...>`, `$b` must be known to be at least as big as `$a`
+- The `ref` of a variable has a lifetime of the 
+    - For a ref local, parameter, field or return of type `ref<$a> T` the lifetime is `$a`
+    - `$heap` for all reference types and fields of reference types
+    - `$local` for everything else
+- An assignment or return is legal when the underlying type conversion is legal
+- Lifetimes of expressions can be made explicit by using cast annotations:
+    - `(T<$a> expr)` the value lifetime is explicitly `$a` for `T<...>`
+    - `ref<$a> (T<$b>)expr` the value lifetime is `$b` for `T<...>` and the ref lifetime is `$a`.
+
+For the purpose of lifetime rules a `ref` is considered part of the type of the expression for purposes of conversions. It is logically represented by converting `ref<$a> T<...>` to `ref<$a, T<...>>` where `$a` is covariant and `T` is invariant. 
```

Comment:
Our other convo about how to represent `ref` in conversions got flagged as outdated cause I updated the text. Here is the updated text so let me know what you think.

Reply:
Yup, looks right to me.

One thing I noted in passing: I think `T` could be covariant in `ref readonly`. But, I think we should actually make the change in the language to allow that universally, not just for lifetimes.

So, for example,

```C#
object o = new();
string s = "";
ref readonly object ro = ref o;
ro = ref s; // Not legal today, but I think it should be
```